src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+                               val (duration, energy) = dispatchEnergy(Int.MaxValue, chargingVehicle, physicalBounds)
+                               chargingVehicle.processChargingCycle(timeBin, energy, duration)
+                             }
+                             ScheduleTrigger(
+                               ChargingTimeOutTrigger(nextTimeBin(timeBin) - 1, vehicle.id, vehicleManager),
+                               self
+                             )
+                         }
+                         .toVector
+                     })
+      )
+
+    case TriggerWithId(ChargingTimeOutTrigger(tick, vehicleId, vehicleManager), triggerId) =>
+      log.debug(s"ChargingTimeOutTrigger for vehicle $vehicleId at $tick")
+      val chargingNetwork = chargingNetworkMap(vehicleManager)
+      chargingNetwork.lookupVehicle(vehicleId) match {
+        case Some(chargingVehicle) => handleEndCharging(tick, chargingVehicle)
+        case _                     => log.debug(s"Vehicle $vehicleId is already disconnected")
+      }
+      sender ! CompletionNotice(triggerId)
+
+    case ChargingPlugRequest(tick, vehicle, vehicleManager) =>
+      log.debug(s"ChargingPlugRequest received for vehicle $vehicle at $tick and stall ${vehicle.stall}")
+      if (vehicle.isBEV | vehicle.isPHEV) {
+        val chargingNetwork = chargingNetworkMap(vehicleManager)
+        // connecting the current vehicle
+        chargingNetwork.connectVehicle(tick, vehicle, sender) match {
+          case chargingVehicle @ ChargingVehicle(vehicle, _, _, station, _, _) if chargingVehicle.status == Waiting =>
+            log.debug(
+              s"Vehicle $vehicle is moved to waiting line at $tick in station $station, with {}/{} vehicles connected and {} in waiting line",
+              station.connectedVehicles.size,
+              station.zone.numChargers,
+              station.waitingLineVehicles.size
+            )
+            log.debug(s"connected vehicles: ${station.connectedVehicles.keys.mkString(",")}")
+            log.debug(s"waiting vehicles: ${station.waitingLineVehicles.keys.mkString(",")}")
+            sender ! WaitingInLine(tick, vehicle.id)
+          case chargingVehicle: ChargingVehicle =>
+            handleStartCharging(tick, chargingVehicle)
+        }
+      } else {
+        sender ! Failure(
+          new RuntimeException(
+            s"$vehicle is not a BEV/PHEV vehicle. Request sent by agent ${sender.path.name}"
+          )
+        )
+      }
+
+    case ChargingUnplugRequest(tick, vehicle, vehicleManager) =>
+      log.debug(s"ChargingUnplugRequest received for vehicle $vehicle from plug ${vehicle.stall} at $tick")
+      val physicalBounds = obtainPowerPhysicalBounds(tick, None)
+      val chargingNetwork = chargingNetworkMap(vehicleManager)
+      chargingNetwork.lookupVehicle(vehicle.id) match {
+        case Some(chargingVehicle) =>
+          val prevStartTime = chargingVehicle.latestChargingCycle.get.startTime
```

Comment:
My head is a bit turned around, so this is just a question to make sure - Does the min/max logic make it so that this handles all scenarios: plugged in but unplugged before the first planning AND plugged in, planned, and then unplugged? I assume it handles the latter, but does it the former?

Reply:
yes it handles both with this
```
          val startTime = Math.min(tick, prevStartTime)
          val endTime = Math.max(tick, prevStartTime)
```